Change context and http to reflect upcoming changes

AnotherCat opened the pull request:
## About this pull request

This pr implements the upcoming UI changes [#2615](https://github.com/discord/discord-api-docs/pull/2615)  
I will leave it as a draft until I am able to test it (discord releases the changes)

## Changes

- Reworked `http` for clarity in context.py
- Removed `respond`, `ack`, and `send_hidden`
- Added `defer`
- Edited `send` to:
  - Check if it's currently deffered, and if so edit it
  - Check if nothing has been sent, and if so respond with type 4
- Set `.message` to the response from either the edit to the initial response or the initial response
- Added a new error `AlreadyResponded` which is raised if `ctx.defer()` is called when the message was already responded to

I have added a warning if the `hidden` value, passed to `send` immeditally after being deffered, if different to the `hidden` value passed to `defer`. 
This is because currently I'm assuming that you can't change the 'state' of the message after responding initially.
I have asked [here](https://github.com/discord/discord-api-docs/pull/2615#issuecomment-783291122)

## Checklist

- [ ] I've checked this pull request runs on `Python 3.6.X`.
- [ ] This fixes something in [Issues](https://github.com/eunwoo1104/discord-py-slash-command/issues).
    - Issue:
- [x] This adds something new.
- [x] There is/are breaking change(s).
- [x] (If required) Relavent documentation has been updated/added.
- [ ] This is not a code change. (README, docs, etc.)

Note: I could not test this code as discord has not released the features yet

AnotherCat commented:
I've: 
Added a new error `AlreadyResponded` which is raised if `ctx.defer()` is called when the message was already responded to

Also have done a few bug fixes
I cannot test deffering messages right now, because all the changes have not rolled out on the discord API yet

eunwoo1104 commented:
Some changes are necessary due to #113 merging.

AnotherCat commented:
Will do them later today
There's also a few issues with ephermical messages i need to fix too

goverfl0w commented:
What's prog. on this? Since we're essentially waiting on the PR merge from Discord to make it official on all send types, OPT_TYPE 4 has this already working now, and early implementation could be considered. Have we made adaptation changes for #113 as of yet?

AnotherCat commented:
Type 5 needs to be changed to deffering the response, which will happen soon (no time frame), because right now if you try to PATCH a deffered response it errors (this was a day ago)
Also need to do a few fixes on the code
I'll merge the new changes on `master` as well
Most likely do these changes later today
Also things like quickstart will need to be updated to

AnotherCat commented:
I've updated references to `ctx.respond` and `ctx.ack`, renamed `migrate_to_109.rst` to `migration.rst`.
I'm not too sure what to do in quickstart because now it's not necessary to call `.defer` then `.send`, I would actually advise against it, and the only references to `.respond` are directly above `.send`. @goverfl0w do you have any ideas? (since you rewrote all of quickstart)

goverfl0w commented:
LTGM, though I recommend we also add doc notes in the quickstart for `.defer()`. Docs are planned to have error handling written down for quickstart FYI, pending the merge.

LordOfPolls commented:
New API is live
I've tested this, and it seems to be working, will add another comment if I find any issues

philip-ulrich commented:
I have no real input into this, but LGTM. Working perfectly once I adjusted some code. :)

AnotherCat commented:
Done some testing, everything looks alright so far
*honestly have no idea about codacy*

eunwoo1104 commented on discord_slash/context.py:
```diff
@@ -20,12 +20,14 @@ class SlashContext:
     :ivar subcommand_name: Subcommand of the command.
     :ivar subcommand_group: Subcommand group of the command.
-    :ivar interaction_id: Interaction ID of the command message.
+    :ivar _interaction_id: Interaction ID of the command message.
```
I don't think `interaction_id` should be protected.

eunwoo1104 commented on discord_slash/context.py:
```diff
@@ -20,12 +20,14 @@ class SlashContext:
+    :ivar _logger: Logger instance.
+    :ivar _deffered: Whether the command is current deffered (loading state)
+    :ivar _deffered_hidden: Internal var to check that state stays the same
+    :ivar _sent: Whether you sent the initial response.
```
Also `sent`, `differed` shouldn't.

AnotherCat commented:
@eunwoo1104 have done those changes, and renamed `sent` to `responded` for better clarity (see convo in server)

eunwoo1104 commented:
Seems good. Merging this pull request!
